src/UserSpaceInstrumentation/Trampoline.cpp

```diff
@@ -0,0 +1,155 @@
+      continue;
+    }
+    std::vector<std::string> addresses = absl::StrSplit(tokens[0], '-');
+    if (addresses.size() != 2) {
+      continue;
+    }
+    uint64_t address_begin = 0;
+    uint64_t address_end = 0;
+    if (!safe_strtou64_base(addresses[0], &address_begin, 16) ||
+        !safe_strtou64_base(addresses[1], &address_end, 16)) {
+      continue;
+    }
+    CHECK(address_begin < address_end);
+    // Join with previous segment ...
+    if (result.back().end == address_begin) {
+      result.back().end = address_end;
+      continue;
+    }
+    // ... or append as new segment.
+    result.emplace_back(address_begin, address_end);
+  }
+  return result;
+}
+
+ErrorMessageOr<AddressRange> FindAddressRangeForTrampoline(
+    const std::vector<AddressRange>& taken_ranges, const AddressRange& code_range, uint64_t size) {
+  constexpr uint64_t kMax32BitOffset = 0x7fffffff;
+  const uint64_t page_size = sysconf(_SC_PAGE_SIZE);
+
+  // Try to fit an interval of length `size` below `code_range`.
+  auto optional_range_index = LowestIntersectingAddressRange(taken_ranges, code_range);
+  if (!optional_range_index) {
+    return ErrorMessage(absl::StrFormat("code_range %#x-%#x is not in taken_ranges.",
+                                        code_range.start, code_range.end));
+  }
+  while (optional_range_index.value() > 0) {
+    // Place directly to the left of the take interval we are in...
+    uint64_t address_trampoline = taken_ranges[optional_range_index.value()].start - size;
+    // ... but round down to the previous page boundary.
+    address_trampoline = (address_trampoline / page_size) * page_size;
+    AddressRange range_trampoline = {address_trampoline, address_trampoline + size};
+    optional_range_index = LowestIntersectingAddressRange(taken_ranges, range_trampoline);
+    if (!optional_range_index) {
+      // We do not intersect any taken interval. Check if we are close enough to code_range:
+      // code_range is above range_trampoline we will need to jmp back and forth in these ranges
+      // with 32 bit offsets. If no distance is greater than 0x7fffffff this is safe.
+      if (code_range.end - range_trampoline.start <= kMax32BitOffset) {
+        return range_trampoline;
+      }
+      // If we are already beyond the close range there is no need to go any further.
+      break;
+    }
+  }
+
+  // Try to fit an interval of length `size` above `code_range`.
+  optional_range_index = HighestIntersectingAddressRange(taken_ranges, code_range);
+  if (!optional_range_index) {
+    return ErrorMessage(absl::StrFormat("code_range %#x-%#x is not in taken_ranges.",
+                                        code_range.start, code_range.end));
+  }
+  do {
+    const uint64_t address_trampoline =
+        ((taken_ranges[optional_range_index.value()].end + (page_size - 1)) / page_size) *
+        page_size;
+    // Check if we ran out of address space.
+    if (address_trampoline >= 0xffffffffffffffff - size) {
```

Comment:
is this UINTPTR_MAX?

Reply:
Probably yes, but that is a coincidence. It is the UINTPTR_MAX of the other process. Currently both are always 64 bit. I hesitate a bit to put UINTPTR_MAX  here since it is misleading in this way. 

I dreamed up my own name "kMax64BitAdress". Not sure if this is better.